### DISCUSSION TITLE
How to increase the loader depth on Connection basic?

### DISCUSSION THREAD

**Discussion by AlexanderPodorov** [sqlalchemy/sqlalchemy#10579] [Usage Questions]
I have a case where I have to eagerly load (using `selectinload`) quite long chain (6-7 levels deep) of related objects. This case has nothing to do with recursion.
What I'm getting is:
```
sys:1: SAWarning: Loader depth for query is excessively deep; caching will be disabled for additional loaders.  Consider using the recursion_depth feature for deeply nested recursive eager loaders.  Use the compiled_cache=None execution option to skip this warning.
```
`recursion_depth` does not apply here, since my loaders are not recursive. If I remove the last level of loading, it emits no warnings.
Is it possible that a recursion is somehow wrongly detected here? If not, how can I increase how many levels deep can I have nested `selectinload`s?

All relationships I'm trying to load are "many-to-one", I changed `selectinload` to `joinedload` for the last two and the warning is no longer emitted, but it looks like a hack for me. Is it the intended behavior?

Probably related to https://github.com/sqlalchemy/sqlalchemy/discussions/10230

Thanks!

**Comment by CaselIT** [comment]
Hi,

This may be a false positive. Can you provide an example of the query you are executing?

Note that that's just a warning, the code will work as expected

**Comment by zzzeek** [comment]
This is not a false positive.  I think people here are mis-reading the warning message, which we can change to improve clarity.

> Loader depth for query is excessively deep; caching will be disabled for additional loaders.  Consider using the recursion_depth feature for deeply nested recursive eager loaders.  Use the compiled_cache=None execution option to skip this warning.


Let me reword here and see if that helps (note this is not what I'm proposing as a change as it's much too wordy):

>  Loader depth for query is excessively deep; caching will be disabled for additional loaders.  **If** this load is a simple recursive load, then consider using the recursion_depth feature for deeply nested recursive eager loaders, which uses an efficient cache key scheme that does not use excessive memory.     Otherwise, if this is simply a very large and deep set of various loaders, setting the compiled_cache=None execution option on this statement will cause all caching to be turned off for this statement, and will additionally cause this warning to not be emitted.

Now what is the big problem with all of that.  It's that setting compiled_cache=None will disable caching for **all** statements from the top level query on down, not just the ones that are past ten levels deep.  Which means we don't have an adequate option here to have it automatically disable statement caching past ten levels deep, while still using caching for the lesser levels, while at the same time not warning.     that said, creating cache keys for loader options that are four, five, six levels deep is still going to use a lot of memory, so perhaps an option like "cache_depth=N" might be appropriate here.

What would be better would be if we can figure out how to get caching for deep loader options to use a top-level cache key, not one that is specific to very deep paths, since a selectinload should not really need to know anything about the upper levels; however there are still a few bits that are cached that **do** need to know the full path, so the problem is not that simple to solve.

**Reply by CaselIT** [reply]
let's try:

> Loader depth for query is excessively deep; caching will be disabled for additional loaders. For recursive eager loader consider using the recursion_depth feature. Use the compiled_cache=None execution option to skip this warning.

**Reply by AlexanderPodorov** [reply]
Thanks a lot for detailed explanation!
I guess the MRE is no longer needed, right?
How many levels deep are supported in practice? I was not able to find that info in docs.
As I mentioned I have 6 levels deep, and it’s the deepest point. Actually I have a tree of loaders with max 6 levels deep.
Does it make sense to switch to joinedload?
Thanks.

**Reply by zzzeek** [reply]
the warning is internally for ten "levels" deep but that counts each mapper plus each relationship as a "level".  so six levels is exactly where it warns.

This patch prints out the level:

```diff
diff --git a/lib/sqlalchemy/orm/context.py b/lib/sqlalchemy/orm/context.py
index 79b43f5fe7..0270bdd009 100644
--- a/lib/sqlalchemy/orm/context.py
+++ b/lib/sqlalchemy/orm/context.py
@@ -512,6 +512,10 @@ class ORMCompileState(AbstractORMCompileState):
                 {"yield_per": load_options._yield_per}
             )
 
+        if (
+            getattr(statement._compile_options, "_current_path", None)
+        ):
+            print(f"DEPTH: {len(statement._compile_options._current_path)}")
         if (
             getattr(statement._compile_options, "_current_path", None)
             and len(statement._compile_options._current_path) > 10

```

we can then run the most minimal test case:

```py
from __future__ import annotations

from sqlalchemy import create_engine
from sqlalchemy import ForeignKey
from sqlalchemy import select
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import relationship
from sqlalchemy.orm import selectinload
from sqlalchemy.orm import Session


class Base(DeclarativeBase):
    pass


class A(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)
    bs: Mapped[list[B]] = relationship()


class B(Base):
    __tablename__ = "b"
    id: Mapped[int] = mapped_column(primary_key=True)
    a_id: Mapped[int] = mapped_column(ForeignKey("a.id"))
    cs: Mapped[list[C]] = relationship()


class C(Base):
    __tablename__ = "c"
    id: Mapped[int] = mapped_column(primary_key=True)
    b_id: Mapped[int] = mapped_column(ForeignKey("b.id"))
    ds: Mapped[list[D]] = relationship()


class D(Base):
    __tablename__ = "d"
    id: Mapped[int] = mapped_column(primary_key=True)
    c_id: Mapped[int] = mapped_column(ForeignKey("c.id"))
    es: Mapped[list[E]] = relationship()


class E(Base):
    __tablename__ = "e"
    id: Mapped[int] = mapped_column(primary_key=True)
    d_id: Mapped[int] = mapped_column(ForeignKey("d.id"))
    fs: Mapped[list[F]] = relationship()


class F(Base):
    __tablename__ = "f"
    id: Mapped[int] = mapped_column(primary_key=True)
    e_id: Mapped[int] = mapped_column(ForeignKey("e.id"))
    gs: Mapped[list[G]] = relationship()


class G(Base):
    __tablename__ = "g"
    id: Mapped[int] = mapped_column(primary_key=True)
    f_id: Mapped[int] = mapped_column(ForeignKey("f.id"))


e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

s = Session(e)

s.add(A(bs=[B(cs=[C(ds=[D(es=[E(fs=[F(gs=[G()])])])])])]))
s.commit()


s.execute(
    select(A).options(
        selectinload(A.bs)
        .selectinload(B.cs)
        .selectinload(C.ds)
        .selectinload(D.es)
        .selectinload(E.fs)
        .selectinload(F.gs)
    )
).all()

```

and see the result printing the depth and the warning:

```
2023-11-04 18:25:16,556 INFO sqlalchemy.engine.Engine SELECT a.id 
FROM a
2023-11-04 18:25:16,556 INFO sqlalchemy.engine.Engine [generated in 0.00016s] ()
DEPTH: 2
2023-11-04 18:25:16,559 INFO sqlalchemy.engine.Engine SELECT b.a_id AS b_a_id, b.id AS b_id 
FROM b 
WHERE b.a_id IN (?)
2023-11-04 18:25:16,559 INFO sqlalchemy.engine.Engine [generated in 0.00022s] (1,)
DEPTH: 4
2023-11-04 18:25:16,562 INFO sqlalchemy.engine.Engine SELECT c.b_id AS c_b_id, c.id AS c_id 
FROM c 
WHERE c.b_id IN (?)
2023-11-04 18:25:16,562 INFO sqlalchemy.engine.Engine [generated in 0.00025s] (1,)
DEPTH: 6
2023-11-04 18:25:16,564 INFO sqlalchemy.engine.Engine SELECT d.c_id AS d_c_id, d.id AS d_id 
FROM d 
WHERE d.c_id IN (?)
2023-11-04 18:25:16,565 INFO sqlalchemy.engine.Engine [generated in 0.00026s] (1,)
DEPTH: 8
2023-11-04 18:25:16,567 INFO sqlalchemy.engine.Engine SELECT e.d_id AS e_d_id, e.id AS e_id 
FROM e 
WHERE e.d_id IN (?)
2023-11-04 18:25:16,567 INFO sqlalchemy.engine.Engine [generated in 0.00019s] (1,)
DEPTH: 10
2023-11-04 18:25:16,569 INFO sqlalchemy.engine.Engine SELECT f.e_id AS f_e_id, f.id AS f_id 
FROM f 
WHERE f.e_id IN (?)
2023-11-04 18:25:16,569 INFO sqlalchemy.engine.Engine [generated in 0.00021s] (1,)
DEPTH: 12
/mnt/photon_home/classic/dev/sqlalchemy/test3.py:82: SAWarning: Loader depth for query is excessively deep; caching will be disabled for additional loaders.  Consider using the recursion_depth feature for deeply nested recursive eager loaders.  Use the compiled_cache=None execution option to skip this warning.
  s.execute(select(A).options(selectinload(A.bs).selectinload(B.cs).selectinload(C.ds).selectinload(D.es).selectinload(E.fs).selectinload(F.gs))).all()
2023-11-04 18:25:16,572 INFO sqlalchemy.engine.Engine SELECT g.f_id AS g_f_id, g.id AS g_id 
FROM g 
WHERE g.f_id IN (?)
2023-11-04 18:25:16,572 INFO sqlalchemy.engine.Engine [caching disabled 0.00028s] (1,)

```

**Reply by AlexanderPodorov** [reply]
Thanks, do you still consider an option to make `DEPTH` configurable?

**Reply by zzzeek** [reply]
yah we can make some kind of option for someone to explicitly state the depth at which they want caching to stop, and then also state the depth in terms of hops between objects

**Reply by AlexanderPodorov** [reply]
That would be great.

**Comment by CaselIT** [comment]
I've updated the wording here https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4961
